src/TargetEventHandlers.js

```diff
@@ -48,40 +46,39 @@ export default class TargetEventHandlers {
       );
     }
 
-    eventHandlers.size += 1;
-    eventHandlers.index += 1;
-    eventHandlers.handlers[eventHandlers.index] = listener;
+    eventHandlers.handlers.push(listener);
 
-    return {
-      target: this.target,
-      eventName,
-      options,
-      index: eventHandlers.index,
-    };
-  }
+    let isSubscribed = true;
+    const unsubscribe = () => {
+      if (!isSubscribed) {
+        return;
+      }
 
-  delete({ eventName, index, options }) {
-    // options has already been normalized at this point.
-    const eventHandlers = this.getEventHandlers(eventName, options);
+      isSubscribed = false;
 
-    if (eventHandlers.size === 0) {
-      // There are no matching event handlers, so no work to be done here.
-      return;
-    }
+      const index = eventHandlers.handlers.indexOf(listener);
+      eventHandlers.handlers.splice(index, 1);
 
-    if (eventHandlers.handlers[index]) {
-      delete eventHandlers.handlers[index];
-      eventHandlers.size -= 1;
-    }
+      if (eventHandlers.handlers.length === 0) {
+        // All event handlers have been removed, so we want to remove the event
+        // listener from the target node.
 
-    if (eventHandlers.size === 0) {
-      this.target.removeEventListener(
-        eventName,
-        eventHandlers.handleEvent,
-        options,
-      );
+        if (this.target) {
```

Comment:
I couldn't think of a good way to write a test for this. If anyone has any ideas, please drop them here.